### ISSUE TITLE
Flushing cache/reset HW to initial state to correctly profile different runs of the same program

### ISSUE THREAD

**Issue by nghielme** [NERSC/timemory-tutorials#3]
As stated in the title: is there a way to reset the state of the HW to "initial" state so that the cache state does not affect the profiling? I notice that different runs of the same program based on the same input data produce different result especially if you consider "input" related metrics.

**Comment by jrmadsen** [comment]
If you could give me a bit more information on the scope of the program and what kind of variance you are seeing I could probably provide more insight but, in general, when doing benchmarking it is generally recommended to discard the first couple iterations or run some initial dummy code to negate some of the energy-saving properties of the electronics which affect performance -- hardware won't go from idle to peak operating frequency immediately bc that can damage the electronics. 

Also, most non-HPC CPUs/GPUs do what Intel calls [turbo boosting](https://en.wikipedia.org/wiki/Intel_Turbo_Boost) which temporarily raise the frequency to a higher value than is sustainable without damaging the electronics. The amplitude and duration of the boost lasts is subject to the heat sink capacity -- if the CPU/GPU is actively cooled and thus can discard the extra heat more effectively, the amplitude/duration is generally higher/longer.

**Comment by jrmadsen** [comment]
I find fibonacci calculations are good ways to warm up the electronics since they are simple to write and can easily scale up:

```cpp
long fibonacci(long n)
{
    return (n < 2) ? n : fibonacci(n-1) + fibonacci(n-2);
}

void warmup(long n = 45)
{
    auto ret = fibonacci(n);
    std::cout << "warmup fibonacci(" << n << ") = " << ret << '\n';
}
```

Just be aware, you will need to _use_ the result of the calculation somewhere (e.g. I printed it above) or else the compiler will optimize away the entire calculation.

**Comment by nghielme** [comment]
Thank you for your exhaustive reply. Just to add more information as you asked: I am trying to estimate how much data is exchanged by my ML algorithm implemented in python so that I can better dimension the communication infrastructure of the same algorithm ported to HW (FPGA for example). I noticed for example that the metrics "IO_in" is (of course) affected if you already run the same program before (because the dataset is probably cached). At the end, I saw 0 IO_in requests if I had already run the script before, that is different with respect to the first time (so probably, I would say, without the dataset in the cache) where I got a value greater than 0 (that depends on the algorithm since it samples randomly from the dataset). I am more interested in the metric read_bytes that seems to correctly identify the accesses to RAM (instead of max_RSS that provides information related to the max amount of memory needed): am I wrong or is it a good metric to consider?
Thank you in advance.

**Comment by jrmadsen** [comment]
So if you want descriptions for each components, run `timemory-avail -d`. Also, I am in the process of adding more documentation to the source so that more detailed descriptions are rendered when the documentation gets generated: [timemory.readthedocs.io/en/develop/components.html](https://timemory.readthedocs.io/en/develop/components.html).

`read_bytes` is generally an I/O metric. On Linux it is generated from reading the `/proc/<PID>/io` file: [source/components/rusage/backends.hpp#L475](https://github.com/NERSC/timemory/blob/9ad64a964a16ab0c31c51b274b19c8312cfee036/source/timemory/components/rusage/backends.hpp#L475). If you want true 'accesses to RAM', you need hardware counters for this. For CPUs, you will want to build timemory with PAPI support and probably `timemory.settings.papi_events = 'PAPI_LD_INS, PAPI_SR_INS'` for the number of load and store instructions. You can explore which HW counters are available and their descriptions via `papi_avail` (provided by PAPI) or `timemory-avail -H -d`. If you want to collecting hardware counters for the GPU, you need to install with CUDA and CUPTI support (both provided by the CUDA toolkit). I don't have support for collecting hardware counters on FPGAs yet but ultimately, that will be added when we (NERSC) start using FPGAs more. Currently, only our advanced technologies group is exploring them. I didn't find an API for collecting those based on a quick google search but if you find one that can collect what you need, create an issue in the main repo and request the feature and I will be able to create a component for it.

**Comment by nghielme** [comment]
Thank you for the precise information you provided me. Regarding the last part related to FPGA: at the moment I am not interested in profiling the algorithm on FPGA, instead I would like to estimate how much data is exchanged between RAM and the CPU while my program is running on my PC so that I can better dimension the communication infrastructure (NOC or BUS) on the FPGA.

**Comment by nghielme** [comment]
After a short search about the `/proc/<PID>/io` file I found this [post](https://stackoverflow.com/questions/3633286/what-do-the-counters-in-proc-pid-io-mean): maybe can be useful also to add the `rchar` and `wchar` bits to the profiling since those metrics are not affected by caching (as stated in the guide). What do you think about that?

**Comment by jrmadsen** [comment]
Hi @nicologhielmetti -- sorry I went on vacation last week. I have seen that post before and it caused to remember that  I implemented reading the `rchar` and `wchar` values as `read_bytes` and `write_bytes` because they appeared to match the values I was seeing on macOS better (which uses and `rusage_info_current.ri_diskio_bytesread` and `rusage_info_current.ri_diskio_byteswritten`) and I didn't really like the idea of the component having different meanings on Linux vs. macOS. Unfortunately, it appears I forgot to update the documentation. 

I am going to try and come up with a unit test for this so that I can try to better ascertain whether `rchar` or `read_bytes` is closer to `rusage_info_current.ri_diskio_bytesread`, etc. and then I will have to try to figure out whether change the meaning of `read_bytes` and `written_bytes` and create new `rchar` and `wchar` components or possibly do something else. I will let you know.
